src/RA_RichPresence.cpp

```diff
@@ -1,266 +1,260 @@
 #include "RA_RichPresence.h"
 
-#include "RA_Core.h"
-#include "RA_GameData.h"
+#include "RA_Defs.h"
 #include "RA_MemValue.h"
 
 RA_RichPresenceInterpreter g_RichPresenceInterpreter;
 
-RA_Lookup::RA_Lookup(const std::string& sDesc)
+RA_RichPresenceInterpreter::Lookup::Lookup(const std::string& sDesc)
     : m_sLookupDescription(sDesc)
 {
 }
 
-const std::string& RA_Lookup::Lookup(ra::DataPos nValue) const
+const std::string& RA_RichPresenceInterpreter::Lookup::GetText(unsigned int nValue) const
 {
-    if (m_lookupData.find(nValue) != m_lookupData.end())
-        return m_lookupData.find(nValue)->second;
+    auto iter = m_mLookupData.find(nValue);
+    if (iter != m_mLookupData.end())
+        return iter->second;
 
-    static const std::string sUnknown = "";
-    return sUnknown;
+    return m_sDefault;
 }
 
-RA_ConditionalDisplayString::RA_ConditionalDisplayString(const char* pBuffer)
+RA_RichPresenceInterpreter::DisplayString::DisplayString()
 {
-    ++pBuffer; // skip first question mark
+    // add an empty core group to ensure Test() always returns true
+    m_conditions.AddGroup(); 
```

Comment:
Won't this cause problems during general use? Setting an error state seems safer (and possibly cheaper) than adding or removing groups.

Reply:
Also, can you clarify the goal of this new function? Is it to display an empty message?

Reply:
This isn't actually an error state. This constructor is for the default display string (no condition). The other constructor is for a conditional display string. After either constructor is called, `InitializeParts` is called to process the display string itself.
```
        for (std::map<std::string, std::string>::const_iterator iter = mDisplayStrings.begin(); iter != mDisplayStrings.end(); ++iter)
        {
            auto& displayString = m_vDisplayStrings.emplace_back(iter->first);
            displayString.InitializeParts(iter->second, mFormats, m_vLookups);
        }

        auto& displayString = m_vDisplayStrings.emplace_back();
        displayString.InitializeParts(sDisplayString, mFormats, m_vLookups);
```
`InitializeParts` could be folded into the constructor.

Reply:
Regarding the "error state": there's actually three states that have to be supported:
1) no condition (always true)
2) invalid condition (always false)
3) valid condition (must be evaluated)

An empty group always evaluates true, a condition with no groups always evaluates false. By doing it this way, the Test logic can just evaluate the condition field without having to check a separate state variable.

Reply:
My mistake, I didn't realize it was a constructor.